### PR TITLE
add a compile-time option to enable 4k page sizes (#52229)

### PR DESCRIPTION
**Pull request by d-netto**
## PR Description

Backports upstream patch to enable 4k page sizes as a build-time option.

## Checklist

Requirements for merging:
- [x] I have opened an issue or PR upstream on JuliaLang/Julia: https://github.com/JuliaLang/julia/pull/52229
- [x] I have removed the `port-to-*` labels that don't apply.
- [x] I have opened a PR on raicode to test these changes: https://github.com/RelationalAI/raicode/pull/16336

### CODE REVIEW

**Comment by d-netto** [comment]
@kpamnany: could you approve it so that we have the basic functionality in place?

Actually flipping the flag to enable it would be another PR, as usual.

File: src/julia_threads.h [hunk 1]
```diff
@@ -160,14 +160,8 @@ typedef struct {
...
     // variables for allocating objects from pools
-#ifdef _P64
-#  define JL_GC_N_POOLS 49
-#elif MAX_ALIGN == 8
-#  define JL_GC_N_POOLS 50
-#else
-#  define JL_GC_N_POOLS 51
-#endif
-    jl_gc_pool_t norm_pools[JL_GC_N_POOLS];
+#define JL_GC_N_MAX_POOLS 51 // conservative. must be kept in sync with `src/julia_internal.h`
```
Comment (kpamnany) [inline src/julia_threads.h:163]: Not sure why this change, or why not have `#ifdef GC_SMALL_PAGE` here also?

**Review by kpamnany — APPROVED** [review]

File: src/julia_threads.h (hunk 1 above)
Comment (d-netto) [reply src/julia_threads.h:163]: All files transitively included from `julia_threads.h` must be marked as `PUBLIC_HEADERS`.

If we added an `#ifdef GC_SMALL_PAGE` then we would need to include `src/options.h` (where this flag is defined) and mark it as a public header as well, which is a bit ugly.
